language/func-api.md

```diff
@@ -539,3 +544,33 @@ called directly from Ruby, then the default applies, if called from the puppet l
 the injected value is used. This may be ideal for unit testing the function since it
 can be tested without using injection and with default for available checkers.
 
+Autoloading
+---
+Functions are autoloaded from files as shown in the table below. The reference `<root>` in the
+shown paths is either the root of the environment (the environment's directory), or the root of a module. The notation `<module name>` means the name of the module without the author part, and `<function_name>` means the leaf name of the function (without name spaces).
+
+| API        | in       | namespace support     | allows top-scope | path |
+| ---        | ---      | ------------------    | ---------------- | --- |
+| 3.x Ruby   | module   | Not supported         | Yes (always top scope) | `<root>/lib/puppet/parser/functions/<function_name>.rb` |
+|            | env      | Not supported         | -   | - 
+| 4.x Ruby   | module   | Yes `<module name>::` | Yes | `<root>/lib/puppet/functions/<module name>/<function_name>.rb` |
+|            | env      | Yes `environment::`   | Yes | `<root>/lib/puppet/functions/environment/<function_name>.rb` `<root>/lib/puppet/functions/<function_name>.rb` |
+| 4.x Puppet | module   | Yes `<module name>::` | No  | `<root>/functions/<function_name>.pp` |
+|            | env      | Yes `environment::`   | No  | `<root>/functions/<function_name>.pp` |
+|            | manifest | Yes (any namspace)    | Yes | in any manifest |
```

Comment:
Yes, you can hurt yourself with that - but not any worse that how you can hurt yourself with the rest. It is also very valuable as it allows you to create functions that override and delegate as a means to overcome version differences. A cool example was adding puppet functions for the `hiera_xxx` functions that delegates to the `lookup` function by placing those functions in the `site.pp`.